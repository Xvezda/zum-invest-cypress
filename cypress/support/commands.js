// ***********************************************
// This example commands.js shows you how to
// create various custom commands and overwrite
// existing commands.
//
// For more comprehensive examples of custom
// commands please read more here:
// https://on.cypress.io/custom-commands
// ***********************************************
//
//
// -- This is a parent command --
// Cypress.Commands.add('login', (email, password) => { ... })
//
//
// -- This is a child command --
// Cypress.Commands.add('drag', { prevSubject: 'element'}, (subject, options) => { ... })
//
//
// -- This is a dual command --
// Cypress.Commands.add('dismiss', { prevSubject: 'optional'}, (subject, options) => { ... })
//
//
// -- This will overwrite an existing command --
// Cypress.Commands.overwrite('visit', (originalFn, url, options) => { ... })

require('cypress-plugin-snapshots/commands');
const { recurse } = require('cypress-recurse');

const { basename } = require('path');
Cypress.Commands.add('fixCypressSpec', function () {
  // https://github.com/cypress-io/cypress/issues/3090#issuecomment-889470707
  const { absoluteFile, relativeFile } = this.test.invocationDetails;
  Cypress.spec = {
    ...Cypress.spec,
    absolute: absoluteFile,
    name: basename(absoluteFile),
    relative: relativeFile,
  };
});

Cypress.Commands.add('stubInvestApi', () => {
  cy.intercept('/api/global', {fixture: 'global'})
    .as('apiGlobal');

  cy.intercept('/api/investment', {fixture: 'investment'})
    .as('apiInvestment');
  cy.intercept('/api/investment/authors*', req => {
      req.reply({fixture: 'investment-authors'});
    })
    .as('apiInvestmentAuthors');
  cy.intercept('/api/investment/authors/*', {fixture: 'investment-author'})
    .as('apiInvestmentAuthor');
  cy.intercept('/api/investment/posts*', req => {
      req.reply({fixture: `investment-home-authors-${req.query.page}`});
    })
    .as('posts');
  cy.intercept('/api/investment/posts/*', {fixture: 'investment-posts'})
    .as('apiInvestmentPosts');
  cy.intercept('/api/investment/home/authors*', req => {
      req.reply({fixture: `investment-home-authors-${req.query.page}`});
    })
    .as('apiAuthors');

  cy.intercept('/api/investment/authors/*/posts/recent**', req => {
      req.reply({fixture: `investment-authors-posts-recent-${req.query.page}`});
    })
    .as('apiAuthorsPostsRecent');

  cy.intercept('/api/domestic/common', {fixture: 'domestic-common'})
    .as('apiDomesticCommon');
  cy.intercept('/api/domestic/home', {fixture: 'domestic-home'})
    .as('apiDomesticHome');
  cy.intercept('/api/domestic/home/meko-chart', {fixture: 'domestic-meko-chart'})
    .as('apiMekoChart');

  cy.intercept('/api/*/home/real-time-news*', req => {
      req.reply({fixture: `real-time-news-${req.query.page}`});
    })
    .as('apiRealTimeNews');


  cy.intercept('/api/suggest*', {fixture: 'search-suggest-zum'})
    .as('apiSuggest');
  cy.intercept('/api/home/category-news*', req => {
      req.reply({fixture: `category-news-${req.query.page}`});
    })
    .as('apiCategoryNews');

  // 정적 컨텐츠를 fixture값으로 대체하기 위해 의도적으로 다른 페이지에서 라우팅하여 이동
  cy.intercept('/api/home', {fixture: 'home'})
    .as('apiHome');

  cy.intercept('/article/info/**', {statusCode: 200})
    .as('apiArticleInfo');
  cy.intercept('/api/news/detail/*', {fixture: 'news-detail'})
    .as('apiNewsDetail');

  cy.intercept('/api/discussion/debate-home/**', {statusCode: 200})
    .as('apiDebateHome');

  cy.intercept('/api/overseas/home', {fixture: 'overseas-home'})
    .as('apiOverseasHome');

  cy.intercept('/api/overseas/home/meko-chart', {fixture: 'overseas-meko-chart'})
    .as('apiOverseasMekoChart');
  cy.intercept('/api/overseas/home/representative-stock*', req => {
      switch (req.query.category) {
        case 'DOW':
          req.reply({fixture: 'overseas-representative-stock-dow'});
          break;
        case 'NASDAQ':
          req.reply({fixture: 'overseas-representative-stock-nasdaq'});
          break;
        default:
          req.destroy();
          break;
      }
    })
    .as('apiOverseasRepresentativeStock');
  cy.intercept('/api/overseas/common', {fixture: 'overseas-common'})
    .as('apiOverseasCommon');

  cy.intercept('https://cmnt.zum.com/article/info/**', {fixture: 'cmnt-article-info'})
    .as('apiCmntArticleInfo');

  cy.intercept('https://cmnt.zum.com/cmnt/article-list/**', {fixture: 'cmnt-article-list'})
    .as('apiCmntArticleList');
});

Cypress.Commands.add(
  'shouldRequestOnScroll',
  (
    alias,
    options,
  ) => {
    const defaultOptions = {
      start: 2,
      count: 2,
      beforeEachScroll: () => {},
      afterEachScroll: () => {},
    };

    const combinedOptions = {
      ...defaultOptions,
      ...options,
    };

    const getPage = page => {
      combinedOptions.beforeEachScroll();
      cy.window()
        .scrollTo('bottom')
        .then(() => {
          combinedOptions.afterEachScroll();
          return cy
            .wait(100)
            .wait(alias);
        })
        .its('request.url')
        .should('contain', `page=${page}`)
        .then(() => {
          if (page <= combinedOptions.count) {
            getPage(page + 1);
          }
        });
    };
    getPage(combinedOptions.start);
  }
);

Cypress.Commands.add(
  'withHidden',
  (selector, callback) => {
    cy.get(selector)
      .then($el => cy.wrap($el.hide()))
      .then($el => {
        callback($el);
        return cy.wrap($el);
      })
      .then($el => cy.wrap($el.show()));
  }
);

Cypress.Commands.add(
  'waitForImage',
  (selector = 'img') => {
    cy.get(selector, {log: false})
      .each($img => {
        cy.wrap($img, {log: false})
          .its('0.naturalWidth', {log: false})
          .should('be.greaterThan', 0);
      });
  }
);

Cypress.Commands.add(
  'ignoreKnownError',
  message => {
    return cy
      .on('uncaught:exception', err => {
        if (message instanceof RegExp) {
          if (message.test(err.message)) {
            return false;
          }
        } else {
          if (err.message.includes(message)) {
            return false;
          }
        }
      }
    );
  }
);

Cypress.Commands.add('login', () => 
  cy.setCookie('_ZIL', '1')  // 로그인 & 로그아웃 표시 버튼
    .setCookie('ZSID', '11111111-2222-3333-4444-555555555555')  // 회원관련 API 요청
    .reload()
);

Cypress.Commands.add('logout', () =>
  cy.clearCookie('_ZIL')
    .clearCookie('ZSID')
    .reload()
);

Cypress.Commands.overwrite('visit', (originalFn, url, options) => {
  const overwrittenOptions = {
    ...(options || {}),
    // 서버측의 응답실패로 인한 테스트 실패의 가능성 최소화
    retryOnStatusCodeFailure: true,
  };
  return originalFn(url, overwrittenOptions);
});