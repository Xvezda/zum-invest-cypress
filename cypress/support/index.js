// ***********************************************************
// This example support/index.js is processed and
// loaded automatically before your test files.
//
// This is a great place to put global configuration and
// behavior that modifies Cypress.
//
// You can change the location of this file or turn off
// automatically serving support files with the
// 'supportFile' configuration option.
//
// You can read more here:
// https://on.cypress.io/configuration
// ***********************************************************

// Import commands.js using ES2015 syntax:
import './commands'

// Alternatively you can use CommonJS syntax:
// require('./commands')

beforeEach(() => {
  cy.fixCypressSpec();

  cy.intercept(/^https?:\/\/thumb\.zumst\.com\/.*/, req => {
    const url = new URL(req.url);
    if (url.pathname.startsWith('/378x241')) {
      req.reply({fixture: '378x241.jpg'});
    } else if (url.pathname.startsWith('/182x77')) {
      req.reply({fixture: '182x77.jpg'});
    } else if (url.pathname.startsWith('/166x116')) {
      req.reply({fixture: '166x116.jpg'});
    } else if (url.pathname.startsWith('/54x34')) {
      req.reply({fixture: '54x34.jpg'});
    } else {
      req.reply({fixture: 'fallback.jpg'});
    }
  });

  cy.intercept('https://pip-thumb.zumst.com/api/v1/**', req => {
    const url = new URL(req.url);
    const width = url.searchParams.get('w');
    const height = url.searchParams.get('h');
    if (width === '880' && height === '495') {
      req.reply({fixture: '880x495.jpg'});
    } else {
      req.reply({fixture: '640x360.jpg'});
    }
  });

  cy.intercept(/^https?:\/\/finance\.zumst\.com\/writing\/.+\.(png|jpe?g|gif)$/, {
    fixture: 'writer.png'
  });
  cy.intercept('https://static.news.zumst.com/images/**', {fixture: '640x360.jpg'});
  cy.intercept('https://finance.zumst.com/content/**', {statusCode: 200});

  cy.intercept('https://zvod.zumst.com/zumvrix/zvod/**', {statusCode: 206});

  cy.intercept('https://cmnt.zum.com/member/login', {fixture: 'member-login'})
    .as('apiMemberLogin');
  cy.fixture('user-info')
    .then(userInfo => {
      cy.intercept('https://userapi.zum.com/getUserInfo*', req => {
        req.reply({
          headers: {
            'Content-Type': 'application/javascript',
          },
          statusCode: 200,
          body: `${req.query.callback}(${JSON.stringify(userInfo)})`,
        });
      });
    });

  cy.on('uncaught:exception', err => {
    if (err.message.includes('kakaoPixel is not defined')) {
      return false;
    }
  });
});

chai.use((chai, utils) => {
  function assertClosest(selector) {
    const $el = utils.flag(this, 'object');
    new chai.Assertion($el.closest(selector)).to.be.exist;
  }

  utils.addMethod(chai.Assertion.prototype, 'ancestors', assertClosest);

  utils.addMethod(chai.Assertion.prototype, 'activated', function () {
    assertClosest.call(this, '.active');
  });
});