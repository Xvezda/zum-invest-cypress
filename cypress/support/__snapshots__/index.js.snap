exports[`zum 투자 홈 > 오늘의 주요뉴스 > 주요뉴스 카드를 클릭하여 투자뉴스 읽기 페이지로 이동할 수 있고 하트를 눌러 좋아요 표시 할 수 있다. #0`] =
{
  "AlternativeInvestmentElement": {
    "authorId": "number",
    "authorName": "string",
    "authorThumbnailUrl": "string",
    "isOriginal": "boolean",
    "leadText": "string",
    "postId": "number",
    "subCategory": "string",
    "title": "string",
    "writeDateTime": "Date"
  },
  "CategoryNews": {
    "all": "DomesticElement[]",
    "coin": "DomesticElement[]",
    "domestic": "DomesticElement[]",
    "esg": "DomesticElement[]",
    "market": "DomesticElement[]",
    "overseas": "DomesticElement[]"
  },
  "DomesticElement": {
    "category": "null | string",
    "id": "string",
    "landingUrl": "string",
    "mediaName": "null | string",
    "registerDateTime": "Date | null",
    "subCategory": "null | string",
    "thumbnail": "string",
    "title": "string"
  },
  "InvestmentContentsNotice": {
    "postId": "number",
    "title": "string",
    "writeDateTime": "Date"
  },
  "MainIndicator": {
    "items": "MainIndicatorItem[]"
  },
  "MainIndicatorItem": {
    "category": "string",
    "currentPrice": "number",
    "id": "string",
    "name": "string",
    "priceChange": "number",
    "rateOfChange": "number",
    "unit": "string"
  },
  "MainNews": {
    "subCategories": "string[]",
    "subNewsItems": "DomesticElement[]",
    "templateCategories": "string[]",
    "templatedNews": "TemplatedNews"
  },
  "News": {
    "items": "DomesticElement[]"
  },
  "Notices": {
    "investmentContentsNotices": "InvestmentContentsNotice[]",
    "performanceSummaryNotices": "PerformanceSummaryNotice[]"
  },
  "PerformanceSummaryNotice": {
    "disclosureDateTime": "Date",
    "stockCode": "string",
    "stockName": "string",
    "title": "string"
  },
  "RealtimeComments": {
    "items": "RealtimeCommentsItem[]"
  },
  "RealtimeCommentsItem": {
    "channelName": "string",
    "content": "string",
    "createDateTime": "Date",
    "key": "string",
    "stockCode": "string",
    "stockIcon": "string",
    "stockName": "string",
    "title": "string"
  },
  "RecentInvestmentContents": {
    "all": "AlternativeInvestmentElement[]",
    "alternativeInvestment": "AlternativeInvestmentElement[]",
    "coin": "AlternativeInvestmentElement[]",
    "domesticStock": "AlternativeInvestmentElement[]",
    "investmentTrend": "AlternativeInvestmentElement[]",
    "overseasStock": "AlternativeInvestmentElement[]"
  },
  "TemplatedNews": {
    "items": "DomesticElement[]",
    "templates": "string[]"
  },
  "TopLevel": {
    "categoryNews": "CategoryNews",
    "mainIndicator": "MainIndicator",
    "mainNews": "MainNews",
    "mostReadNews": "News",
    "notices": "Notices",
    "realtimeComments": "RealtimeComments",
    "recentInvestmentContents": "RecentInvestmentContents",
    "todayMarketNews": "News",
    "videos": "Videos"
  },
  "Videos": {
    "invenId": "string",
    "items": "VideosItem[]"
  },
  "VideosItem": {
    "body": "string",
    "categoryCode": "string",
    "cpCode": "string",
    "cpName": "string",
    "createDateTime": "Date",
    "enabled": "boolean",
    "episode": "number",
    "id": "string",
    "playTime": "number",
    "publishDateTime": "Date",
    "sectionCode": "string",
    "targetAge": "number",
    "targetNation": "string",
    "title": "string",
    "updateDateTime": "Date",
    "vodImage": "string"
  }
};

exports[`zum 투자 홈 > 오늘의 주요뉴스 > 주요뉴스 카드를 클릭하여 투자뉴스 읽기 페이지로 이동할 수 있고 하트를 눌러 좋아요 표시 할 수 있다. #1`] =
{
  "AlternativeInvestmentElement": {
    "authorId": "number",
    "authorName": "string",
    "authorThumbnailUrl": "string",
    "isOriginal": "boolean",
    "leadText": "string",
    "postId": "number",
    "subCategory": "string",
    "title": "string",
    "writeDateTime": "Date"
  },
  "CategoryNews": {
    "all": "DomesticElement[]",
    "coin": "DomesticElement[]",
    "domestic": "DomesticElement[]",
    "esg": "DomesticElement[]",
    "market": "DomesticElement[]",
    "overseas": "DomesticElement[]"
  },
  "DomesticElement": {
    "category": "null | string",
    "id": "string",
    "landingUrl": "string",
    "mediaName": "null | string",
    "registerDateTime": "Date | null",
    "subCategory": "null | string",
    "thumbnail": "null | string",
    "title": "string"
  },
  "InvestmentContentsNotice": {
    "postId": "number",
    "title": "string",
    "writeDateTime": "Date"
  },
  "MainIndicator": {
    "items": "MainIndicatorItem[]"
  },
  "MainIndicatorItem": {
    "category": "string",
    "currentPrice": "number",
    "id": "string",
    "name": "string",
    "priceChange": "number",
    "rateOfChange": "number",
    "unit": "string"
  },
  "MainNews": {
    "subCategories": "string[]",
    "subNewsItems": "DomesticElement[]",
    "templateCategories": "string[]",
    "templatedNews": "TemplatedNews"
  },
  "News": {
    "items": "DomesticElement[]"
  },
  "Notices": {
    "investmentContentsNotices": "InvestmentContentsNotice[]",
    "performanceSummaryNotices": "PerformanceSummaryNotice[]"
  },
  "PerformanceSummaryNotice": {
    "disclosureDateTime": "Date",
    "stockCode": "string",
    "stockName": "string",
    "title": "string"
  },
  "RealtimeComments": {
    "items": "RealtimeCommentsItem[]"
  },
  "RealtimeCommentsItem": {
    "channelName": "string",
    "content": "string",
    "createDateTime": "Date",
    "key": "string",
    "stockCode": "string",
    "stockIcon": "string",
    "stockName": "string",
    "title": "string"
  },
  "RecentInvestmentContents": {
    "all": "AlternativeInvestmentElement[]",
    "alternativeInvestment": "AlternativeInvestmentElement[]",
    "coin": "AlternativeInvestmentElement[]",
    "domesticStock": "AlternativeInvestmentElement[]",
    "investmentTrend": "AlternativeInvestmentElement[]",
    "overseasStock": "AlternativeInvestmentElement[]"
  },
  "TemplatedNews": {
    "items": "DomesticElement[]",
    "templates": "string[]"
  },
  "TopLevel": {
    "categoryNews": "CategoryNews",
    "mainIndicator": "MainIndicator",
    "mainNews": "MainNews",
    "mostReadNews": "News",
    "notices": "Notices",
    "realtimeComments": "RealtimeComments",
    "recentInvestmentContents": "RecentInvestmentContents",
    "todayMarketNews": "News",
    "videos": "Videos"
  },
  "Videos": {
    "invenId": "string",
    "items": "VideosItem[]"
  },
  "VideosItem": {
    "body": "string",
    "categoryCode": "string",
    "cpCode": "string",
    "cpName": "string",
    "createDateTime": "Date",
    "enabled": "boolean",
    "episode": "number",
    "id": "string",
    "playTime": "number",
    "publishDateTime": "Date",
    "sectionCode": "string",
    "targetAge": "number",
    "targetNation": "string",
    "title": "string",
    "updateDateTime": "Date",
    "vodImage": "string"
  }
};

exports[`zum 투자 홈 > 오늘의 주요뉴스 > 주요뉴스 카드를 클릭하여 투자뉴스 읽기 페이지로 이동할 수 있고 하트를 눌러 좋아요 표시 할 수 있다. #2`] =
{
  "AlternativeInvestmentElement": {
    "authorId": "number",
    "authorName": "string",
    "authorThumbnailUrl": "string",
    "isOriginal": "boolean",
    "leadText": "string",
    "postId": "number",
    "subCategory": "string",
    "title": "string",
    "writeDateTime": "Date"
  },
  "CategoryNews": {
    "all": "DomesticElement[]",
    "coin": "DomesticElement[]",
    "domestic": "DomesticElement[]",
    "esg": "DomesticElement[]",
    "market": "DomesticElement[]",
    "overseas": "DomesticElement[]"
  },
  "DomesticElement": {
    "category": "null | string",
    "id": "string",
    "landingUrl": "string",
    "mediaName": "null | string",
    "registerDateTime": "Date | null",
    "subCategory": "null | string",
    "thumbnail": "string",
    "title": "string"
  },
  "InvestmentContentsNotice": {
    "postId": "number",
    "title": "string",
    "writeDateTime": "Date"
  },
  "MainIndicator": {
    "items": "MainIndicatorItem[]"
  },
  "MainIndicatorItem": {
    "category": "string",
    "currentPrice": "number",
    "id": "string",
    "name": "string",
    "priceChange": "number",
    "rateOfChange": "number",
    "unit": "string"
  },
  "MainNews": {
    "subCategories": "string[]",
    "subNewsItems": "DomesticElement[]",
    "templateCategories": "string[]",
    "templatedNews": "TemplatedNews"
  },
  "News": {
    "items": "DomesticElement[]"
  },
  "Notices": {
    "investmentContentsNotices": "InvestmentContentsNotice[]",
    "performanceSummaryNotices": "PerformanceSummaryNotice[]"
  },
  "PerformanceSummaryNotice": {
    "disclosureDateTime": "Date",
    "stockCode": "string",
    "stockName": "string",
    "title": "string"
  },
  "RealtimeComments": {
    "items": "RealtimeCommentsItem[]"
  },
  "RealtimeCommentsItem": {
    "channelName": "string",
    "content": "string",
    "createDateTime": "Date",
    "key": "string",
    "stockCode": "string",
    "stockIcon": "string",
    "stockName": "string",
    "title": "string"
  },
  "RecentInvestmentContents": {
    "all": "AlternativeInvestmentElement[]",
    "alternativeInvestment": "AlternativeInvestmentElement[]",
    "coin": "AlternativeInvestmentElement[]",
    "domesticStock": "AlternativeInvestmentElement[]",
    "investmentTrend": "AlternativeInvestmentElement[]",
    "overseasStock": "AlternativeInvestmentElement[]"
  },
  "TemplatedNews": {
    "items": "DomesticElement[]",
    "templates": "string[]"
  },
  "TopLevel": {
    "categoryNews": "CategoryNews",
    "mainIndicator": "MainIndicator",
    "mainNews": "MainNews",
    "mostReadNews": "News",
    "notices": "Notices",
    "realtimeComments": "RealtimeComments",
    "recentInvestmentContents": "RecentInvestmentContents",
    "todayMarketNews": "News",
    "videos": "Videos"
  },
  "Videos": {
    "invenId": "string",
    "items": "VideosItem[]"
  },
  "VideosItem": {
    "body": "string",
    "categoryCode": "string",
    "cpCode": "string",
    "cpName": "string",
    "createDateTime": "Date",
    "enabled": "boolean",
    "episode": "number",
    "id": "string",
    "playTime": "number",
    "publishDateTime": "Date",
    "sectionCode": "string",
    "targetAge": "number",
    "targetNation": "string",
    "title": "string",
    "updateDateTime": "Date",
    "vodImage": "string"
  }
};

exports[`zum 투자 홈 > 분야별 실시간 뉴스 > 스크롤을 내리면 다음 페이지를 불러온다. #0`] =
{
  "Item": {
    "category": "string",
    "id": "string",
    "landingUrl": "string",
    "leadText": "string",
    "mediaName": "string",
    "registerDateTime": "Date",
    "subCategory": "string",
    "thumbnail": "string",
    "title": "string"
  },
  "TopLevel": {
    "items": "Item[]",
    "totalCount": "number"
  }
};

exports[`해외증시 > 해외 대표 종목 > 기간 버튼을 눌러 차트를 바꾸고 리스트를 클릭하여 뉴스를 바꿀 수 있다. #0`] =
{
  "DebateItem": {
    "agreeCount": "number",
    "agreePercent": "number",
    "articles": "null",
    "content": "string",
    "createdDateTime": "Date",
    "disagreeCount": "number",
    "disagreePercent": "number",
    "endedDateTime": "Date",
    "firstOpinion": "string",
    "id": "number",
    "numberOfParticipants": "number",
    "popularityEnabledDateTime": "Date",
    "secondOpinion": "string",
    "startedDateTime": "Date",
    "theme": "string",
    "title": "string"
  },
  "Dow": {
    "articleId": "null | string",
    "articleTitle": "null | string",
    "stockCode": "string",
    "stockName": "string"
  },
  "LiveNew": {
    "id": "string",
    "title": "string"
  },
  "MainNews": {
    "dow": "Dow[]",
    "nasdaq": "Dow[]"
  },
  "OfMainIndex": {
    "currentPrice": "number",
    "financeCategory": "string",
    "id": "string",
    "name": "string",
    "priceChange": "number",
    "rateOfChange": "number",
    "updateDateTime": "Date",
    "yearlyBottom": "number",
    "yearlyTop": "number"
  },
  "RealTimeNews": {
    "items": "RepresentativeNew[]",
    "totalCount": "number"
  },
  "RecentInvestmentContent": {
    "authorId": "number",
    "authorName": "string",
    "authorThumbnailUrl": "string",
    "isOriginal": "boolean",
    "leadText": "string",
    "postId": "number",
    "subCategory": "string",
    "title": "string",
    "writeDateTime": "Date"
  },
  "RepresentativeNew": {
    "id": "string",
    "landingUrl": "null | string",
    "leadText": "string",
    "mediaName": "string",
    "registerDateTime": "Date",
    "thumbnail": "null | string",
    "title": "string"
  },
  "RepresentativeStock": {
    "article": "RepresentativeNew",
    "stock": "Stock"
  },
  "Stock": {
    "currentPrice": "number",
    "financeCategory": "string",
    "highPrice": "number",
    "id": "string",
    "lowPrice": "number",
    "name": "string",
    "preClosingPrice": "number",
    "priceChange": "number",
    "rateOfChange": "number",
    "symbol": "string",
    "updateDateTime": "Date"
  },
  "TopLevel": {
    "countryIndexOfMainIndex": "OfMainIndex[]",
    "debateItems": "DebateItem[]",
    "indicatorOfMainIndex": "OfMainIndex[]",
    "liveNews": "LiveNew[]",
    "mainNews": "MainNews",
    "realTimeNews": "RealTimeNews",
    "recentInvestmentContents": "RecentInvestmentContent[]",
    "representativeNews": "RepresentativeNew[]",
    "representativeStock": "RepresentativeStock[]"
  }
};
